src/main/java/org/apache/logging/log4j/ext/kusto/KustoClientInstance.java

```diff
@@ -0,0 +1,185 @@
+        }
+        LOGGER.info("Using database : {}  & table {} to ingest logs", kustoLog4jConfig.dbName,
+                kustoLog4jConfig.tableName);
+        ingestionProperties = new IngestionProperties(kustoLog4jConfig.dbName, kustoLog4jConfig.tableName);
+        ingestionProperties.setFlushImmediately(kustoLog4jConfig.flushImmediately);
+        if (kustoLog4jConfig.logTableMapping != null && kustoLog4jConfig.mappingType != null
+                && !"".equals(kustoLog4jConfig.logTableMapping.trim()) &&
+                !"".equals(kustoLog4jConfig.mappingType.trim())) {
+            LOGGER.error("Using mapping {}  of type  {} ", kustoLog4jConfig.logTableMapping,
+                    kustoLog4jConfig.mappingType);
+            IngestionMapping.IngestionMappingKind mappingType = "json".equalsIgnoreCase(kustoLog4jConfig.mappingType) ? JSON : CSV;
+            ingestionProperties.getIngestionMapping()
+                    .setIngestionMappingReference(kustoLog4jConfig.logTableMapping, mappingType);
+        }
+    }
+
+    /**
+     * The instance is created with the config created
+     *
+     * @param kustoLog4jConfig The config as passed by the application
+     * @return The KustoClientInstance that will be operated on in the rest of the application
+     * @throws URISyntaxException When the URI passed is invalid
+     */
+    static KustoClientInstance getInstance(KustoLog4jConfig kustoLog4jConfig) throws URISyntaxException {
+        KustoClientInstance result = ATOMIC_INSTANCE.get();
+        if (result != null) {
+            return result;
+        }
+        synchronized (KustoClientInstance.class) {
+            if (ATOMIC_INSTANCE.get() == null) {
+                ATOMIC_INSTANCE.set(new KustoClientInstance(kustoLog4jConfig));
+            }
+            return ATOMIC_INSTANCE.get();
+        }
+    }
+
+    /**
+     * The instance that is created with the config is returned for use in parts of the application
+     *
+     * @return The KustoClientInstance that will be operated on in the rest of the application
+     */
+    static KustoClientInstance getInstance() {
+        return ATOMIC_INSTANCE.get();
+    }
+
+    void ingestRolledFile(String filePath) {
+        try {
+            ingestionRetry.executeCheckedSupplier(() -> ingestLogs(filePath));
+        } catch (Throwable e) {
+            backOutFile(filePath);
+        }
+    }
+
+    IngestionResult ingestLogs(String filePath) throws IngestionClientException, IngestionServiceException,
+            IOException {
+        try (InputStream inputStream = Files.newInputStream(Paths.get(filePath))) {
+            StreamSourceInfo streamSourceInfo = new StreamSourceInfo(inputStream);
+            return ingestClient.ingestFromStream(streamSourceInfo, ingestionProperties);
+        }
+    }
+
+    /**
+     * In the case the ingestion fails after retries , it is moved to a backout directory that is in the same folder as the path

```

Comment:
```suggestion
     * In case, if ingestion fails after retries, it is moved to a backout directory that is in the same folder as the path

```

Reply:
@asaharn Thanks so much for the review , fixed most of the comments where applicable. Kindly re-review when you get a chance